### PR TITLE
Add upper limit for the program size to prevent tsserver from crashing

### LINKED ISSUE #7444: tsserver crashes with large projects
I am unsure if this is caused by TS eagerly loading [all syntax trees](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/program.ts#L751) or by the tsserver eagerly loading [all symbols](https://github.com/Microsoft/TypeScript/blob/master/src/server/editorServices.ts#L413) (which btw seems crazy to me) but I think the tsserver should check how large a project is (or how much memory is already [used](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/sys.ts#L23)) and fail gracefully. 

I have a JavaScript project with little code but a lots of dependencies like `node_modules` and `bower_components` as well as files that get generated into a `temp` folder (on server start). I am aware of the exclude-property and the [default](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/commandLineParser.ts#L569) [excludes](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/commandLineParser.ts#L591), but the challenge is to provide guidance to the user to use them.  I have no chance to ask tsserver how many files are included by a project because it crashes already with the first `open-file` request. 

A diagnostics message ala `too many files (foo/jsconfig.json, files: [...]), ` would be much appreciated because otherwise we are going to duplicate a lot of logic on our (VS Code) end for such checks (and the server will still crashes).

### LINKED ISSUE #8212: JS error highlighting stops working in some cases
_From @nrkn on April 20, 2016 4:46_

- VSCode Version: 1.0.0
- OS Version: Windows 10 Pro

I have no extensions installed. node.js v5.10.1

Error highlighting has stopped working in our node.js app - I was able to reduce it to a very minimal test case:

A single index.js containing an erroneous line so you can see if error highlighting is on
A single subfolder containing a package.json with grunt and grunt-browserify packages and a node_modules folder created by running npm install against the package.json
Standard vscode files - .vscode/launch.json and jsconfig.json

Steps to Reproduce:

1. [Clone this repo](https://github.com/nrkn/vscode-error-min/) or [recreate structure above from this gist](https://gist.github.com/nrkn/da6bde225b114d9b8d9f7cf66772c0c9)
2. Navigate to ./sub and run `npm install`
3. Open main folder in VSCode
4. View ./index.js - there is no error highlighting

I have reproduced this on 3 machines

Almost any change to this minimum case, like removing the node_modules folder, removing `use strict` from the index.js etc., removing the vscode files etc. prevents the bug from being triggered, and error highlighting starts working again, however our app is not a minimal test case so knowing that isn't much help to us!

_Copied from original issue: Microsoft/vscode#5530_

### PR DESCRIPTION
**Pull request by zhengbli**
Fixes #7444 
For experimental purpose, the upper limit was set to 35 Mb (as suggested in https://github.com/Microsoft/TypeScript/issues/7444#issuecomment-194735638), which might be changed to a better value after further tests. 

Per our discussion in #7444, the next step is to add a compiler switch (something like disableSizeLimit) to disable this check for a project. Though I will do that after the current size-checking logic was reviewed. 

Edit: also fix #8212

### CODE REVIEW

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -2824,5 +2824,9 @@
     "Unknown typing option '{0}'.": {
         "category": "Error",
         "code": 17010
+    },
+    "Too many javascript files in the project. Consider add to the `exclude` list in the config file.": {
```
Comment (billti) [inline src/compiler/diagnosticMessages.json]: Grammar: "Consider adding to...". Or perhaps just reworded as "Use the 'exclude' setting in project configuration to limit included source folders".

Note: Do we predicate this message on the user *not* having listed the files to compile? If we are not scanning directories anyway (i.e. there is a list of files given), then this message would be misleading (as `exclude` can't be specified - at least until the globbing change comes in).

Also, capitalize as "JavaScript", not "javascript", and don't use backticks. This isn't markdown :-p

File: src/compiler/program.ts [hunk 2]
```diff
@@ -748,7 +748,21 @@ namespace ts {
...
 
         if (!tryReuseStructureFromOldProgram()) {
-            forEach(rootNames, name => processRootFile(name, /*isDefaultLib*/ false));
+            let programSize = 0;
+            for (const name of rootNames) {
+                const path = toPath(name, currentDirectory, getCanonicalFileName);
+                if (programSize <= maxProgramSize) {
+                    processRootFile(name, /*isDefaultLib*/ false);
+                    if (!hasTypeScriptFileExtension(name) && filesByName.get(path)) {
+                        programSize += filesByName.get(path).text.length;
```
Comment (billti) [inline src/compiler/program.ts]: Looking in `findSourceFile` (https://github.com/Microsoft/TypeScript/blob/master/src/compiler/program.ts#L1466) I believe the filename can be present even if no file was found/read, thus you might want to check the `text` value is present before trying to access its `length`.

File: src/compiler/program.ts [hunk 3]
```diff
@@ -748,7 +748,21 @@ namespace ts {
...
+                    }
+                }
+                else {
+                    programDiagnostics.add(createCompilerDiagnostic(Diagnostics.Too_many_javascript_files_in_the_project_Consider_add_to_the_exclude_list_in_the_config_file));
```
Comment (billti) [inline src/compiler/program.ts]: I thought we were going to indicate the path to the file being processed when the limit was exceeded (as this is likely in whatever folder you should have been excluding)?

File: src/compiler/sys.ts [hunk 4]
```diff
@@ -500,7 +500,10 @@ namespace ts {
                     for (const current of files) {
                         const name = combinePaths(path, current);
                         if (!contains(exclude, getCanonicalPath(name))) {
-                            const stat = _fs.statSync(name);
+                            // fs.statSync would throw an exception if the file is a symlink
+                            // whose linked file doesn't exist. fs.lstatSync would return a stat
+                            // object for the symlink file itself in this case
+                            const stat = _fs.lstatSync(name);
```
Comment (billti) [inline src/compiler/sys.ts]: This seems like a non-trivial change with a lot of nuance (for example, this means we won't traverse into directories now if they are created as symbolic links, as `isDirectory()` returns `false` on a `lstat` where it used to return `true` on a `stat`.

How is this part of setting a limit on file size when loading? Seems like this is a different issue and should be coded/evaluated separately.

File: src/compiler/program.ts [hunk 5]
```diff
@@ -748,7 +748,21 @@ namespace ts {
         }
 
         if (!tryReuseStructureFromOldProgram()) {
-            forEach(rootNames, name => processRootFile(name, /*isDefaultLib*/ false));
+            let programSize = 0;
+            for (const name of rootNames) {
+                const path = toPath(name, currentDirectory, getCanonicalFileName);
+                if (programSize <= maxProgramSize) {
```
Comment (billti) [inline src/compiler/program.ts]: We [discussed a compiler option](https://github.com/Microsoft/TypeScript/issues/7444#issuecomment-194491024) to disable the size check for cases where the project is legitimately large. I don't see that option in this code review.

File: src/server/editorServices.ts [hunk 6]
```diff
@@ -1217,13 +1217,35 @@ namespace ts.server {
             }
             else {
                 const project = this.createProject(configFilename, projectOptions);
+                let programSize = 0;
+
+                // As the project openning might not be complete if there are too many files,
+                // therefore to surface the diagnostics we need to make sure the given client file is opened.
+                if (clientFileName) {
+                    const currentClientFileInfo = this.openFile(clientFileName, /*openedByClient*/ true);
```
Comment (billti) [inline src/server/editorServices.ts]: Do you not need to worry about the `this.host.fileExists` check that used to be performed for all files? (i.e. is the clientFileName guaranteed to exit)?

File: src/server/editorServices.ts [hunk 7]
```diff
@@ -1217,13 +1217,35 @@ namespace ts.server {
...
+
                 for (const rootFilename of projectOptions.files) {
-                    if (this.host.fileExists(rootFilename)) {
-                        const info = this.openFile(rootFilename, /*openedByClient*/ clientFileName == rootFilename);
-                        project.addRoot(info);
+                    if (rootFilename === clientFileName) {
+                        continue;
+                    }
+
+                    if (programSize <= maxProgramSize) {
```
Comment (billti) [inline src/server/editorServices.ts]: I'm trying to understand why we need to repeat this check for the size of the content loaded here as well as in `createProgram`. It seems dangerous to have similar logic in both places they may behave slightly differently (or diverge unknowingly over time). What is the effect of **not** having this check here?

**Comment by billti** [comment]
Added some comments. Note this pull request should be against `release-1.8` also if VSCode wants it for their release, not against `master`. Thanks!

File: src/compiler/diagnosticMessages.json (hunk 1 above)
Comment (zhengbli) [inline src/compiler/diagnosticMessages.json]: I agree that we should mention both cases, with or without the "exclude" list. Does the following sound better:

"Too many JavaScript files in the project. Use an exact 'files' list, or use the 'exclude' setting  in project configuration to limit included source folders."

File: src/compiler/program.ts (hunk 2 above)
Comment (zhengbli) [inline src/compiler/program.ts]: Good catch!

File: src/compiler/diagnosticMessages.json (hunk 1 above)
Comment (billti) [inline src/compiler/diagnosticMessages.json]: I like it :-)

File: src/compiler/sys.ts (hunk 4 above)
Comment (zhengbli) [inline src/compiler/sys.ts]: This was an issue that caused tsserver to crash when editing the repro sample. The ember server generates tons of symlink files in the `tmp` folder, and then deletes the linked files after the server shuts down. Then these symlink files became invalid ones, which caused `fs.statSync` to throw exceptions when we try to scan the folder.

The alternative solution was to try `fs.statSync` first, if an exception was thrown just skip this file.

File: src/server/editorServices.ts (hunk 7 above)
Comment (zhengbli) [inline src/server/editorServices.ts]: Because tsserver [opens up files](https://github.com/Microsoft/TypeScript/blob/master/src/server/editorServices.ts#L1222) before [calling the `createProgram` via `compilerService`](https://github.com/Microsoft/TypeScript/blob/master/src/server/editorServices.ts#L1229). And its [`openFile`](https://github.com/Microsoft/TypeScript/blob/master/src/server/editorServices.ts#L991) function does some pretty complex logic for file version management. In the `createProgram`, the compiler calls `findSourceFile` function, which will load the file if it found the file wasn't loaded. Therefore I need to set the limit on both sides. I didn't find an more optimized solution to unify them yet.

File: src/compiler/utilities.ts [hunk 8]
```diff
@@ -2866,4 +2870,6 @@ namespace ts {
     export function isParameterPropertyDeclaration(node: ParameterDeclaration): boolean {
         return node.flags & NodeFlags.AccessibilityModifier && node.parent.kind === SyntaxKind.Constructor && isClassLike(node.parent.parent);
     }
+
+    export const maxProgramSize = 20 * 1024 * 1024;
```
Comment (zhengbli) [inline src/compiler/utilities.ts]: Changed the upper limit from 35M to 20M, as on my machine even 35M would still run out of memory

File: src/compiler/commandLineParser.ts [hunk 9]
```diff
@@ -332,6 +332,11 @@ namespace ts {
             name: "noImplicitUseStrict",
             type: "boolean",
             description: Diagnostics.Do_not_emit_use_strict_directives_in_module_output
+        },
+        {
+            name: "disableSizeLimit",
+            type: "boolean",
+            description: Diagnostics.Disable_the_upper_limit_for_the_total_file_size_of_a_project
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: remove the description, so that it does not show in the help.

File: src/compiler/program.ts [hunk 10]
```diff
@@ -742,13 +742,41 @@ namespace ts {
...
-                (oldOptions.allowJs !== options.allowJs)) {
+                (oldOptions.allowJs !== options.allowJs) ||
+                (oldOptions.disableSizeLimit !== options.disableSizeLimit)) {
                 oldProgram = undefined;
             }
         }
 
         if (!tryReuseStructureFromOldProgram()) {
-            forEach(rootNames, name => processRootFile(name, /*isDefaultLib*/ false));
+            if (options.disableSizeLimit === true) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: if (options.disableSizeLimit)

File: src/compiler/diagnosticMessages.json [hunk 11]
```diff
@@ -2824,5 +2827,9 @@
     "Unknown typing option '{0}'.": {
         "category": "Error",
         "code": 17010
+    },
+    "Too many JavaScript files in the project. Use an exact 'files' list, or use the 'exclude' setting in project configuration to limit included source folders. The likely folder to exclude is '{0}'. To disable the project size limit, set the 'disableSizeLimit' compiler option to 'true'": {
```
Comment (mhegazy) [inline src/compiler/diagnosticMessages.json]: consider updating the error message to:

`Too many JavaScript files in the project. Consider specifying the 'exclude' setting in project configuration to limit included source folders. 
The likely folder to exclude is '{0}'. 
To disable the project size limit, set the 'disableSizeLimit' compiler option to 'true'`

**Comment by zhengbli** [comment]
@billti @mhegazy @vladima Please review the updated code, I'll try to merge into 1.8 once this PR got thumbs up. Thanks!

File: src/compiler/program.ts [hunk 12]
```diff
@@ -1452,6 +1454,27 @@ namespace ts {
                 return file;
             }
 
+            if (!options.disableSizeLimit) {
+                if (programSizeForNonTsFiles === -1) {
+                    return;
```
Comment (mhegazy) [inline src/compiler/program.ts]: nit. `return undefined;`

File: src/compiler/program.ts [hunk 13]
```diff
@@ -1452,6 +1454,27 @@ namespace ts {
...
+                    // if the commonSourceDirectory is "/src/", and the last processed path was "/src/node_modules/a/b.js",
+                    // we should show in the error message "/src/node_modules/".
+                    const commonSourceDirectory = getCommonSourceDirectory();
+                    let rootLevelDirectory = path.substring(0, Math.max(commonSourceDirectory.length, path.indexOf(directorySeparator, commonSourceDirectory.length)));
+                    if (rootLevelDirectory[rootLevelDirectory.length - 1] !== directorySeparator) {
+                        rootLevelDirectory += directorySeparator;
+                    }
+                    programDiagnostics.add(createCompilerDiagnostic(Diagnostics.Too_many_JavaScript_files_in_the_project_Consider_specifying_the_exclude_setting_in_project_configuration_to_limit_included_source_folders_The_likely_folder_to_exclude_is_0_To_disable_the_project_size_limit_set_the_disableSizeLimit_compiler_option_to_true, rootLevelDirectory));
+                    programSizeForNonTsFiles = -1;
+                    return;
```
Comment (mhegazy) [inline src/compiler/program.ts]: and this too.

File: src/compiler/utilities.ts [hunk 14]
```diff
@@ -2866,4 +2870,6 @@ namespace ts {
     export function isParameterPropertyDeclaration(node: ParameterDeclaration): boolean {
         return node.flags & NodeFlags.AccessibilityModifier && node.parent.kind === SyntaxKind.Constructor && isClassLike(node.parent.parent);
     }
+
+    export const maxProgramSizeForNonTsFiles = 20 * 1024 * 1024;
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: i would put this in program. and no need to export it. we only use it there.

File: src/server/editorServices.ts [hunk 15]
```diff
@@ -1217,10 +1217,43 @@ namespace ts.server {
             }
             else {
                 const project = this.createProject(configFilename, projectOptions);
+                let programSizeForNonTsFiles = 0;
+
+                // As the project openning might not be complete if there are too many files,
+                // therefore to surface the diagnostics we need to make sure the given client file is opened.
+                if (clientFileName) {
```
Comment (mhegazy) [inline src/server/editorServices.ts]: i thought we said we did not need this anymore? is this not the case?

Comment (zhengbli) [inline src/server/editorServices.ts]: The file reading here doesn't have an upper limit, therefore it could be much more than 20M. In the repro case it is around 100M, and I noticed server crashes. So I kept the limit here as well.

File: src/compiler/utilities.ts (hunk 14 above)
Comment (zhengbli) [inline src/compiler/utilities.ts]: For now it is shared between the server and the program.ts though

**Comment by zhengbli** [comment]
With the new updates, the response time for the first "ProjectInfo" request was reduced to around 5s from 40+s originally, due to avoiding creating programs for the project and speeding up in the file traversing process.

File: src/compiler/commandLineParser.ts [hunk 16]
```diff
@@ -633,10 +637,22 @@ namespace ts {
...
+                    addRange(potentialFiles, host.readDirectoryWithMultipleExtensions(basePath, supportedExtensions, exclude));
+                }
+                else {
+                    for (const extension of supportedExtensions) {
+                        addRange(potentialFiles, host.readDirectory(basePath, extension, exclude));
+                    }
+                }
                 // Get files of supported extensions in their order of resolution
                 for (const extension of supportedExtensions) {
-                    const filesInDirWithExtension = host.readDirectory(basePath, extension, exclude);
```
Comment (billti) [inline src/compiler/commandLineParser.ts]: We were re-reading the directory for every file extension in the list? Crazy!! Good find :smile:

**Comment by billti** [comment]
Looks good at first read. We should add the same logic on the VS host side also to improve Salsa experience there.

**Comment by zhengbli** [comment]
Ping @mhegazy

File: src/compiler/program.ts [hunk 17]
```diff
@@ -1463,6 +1477,25 @@ namespace ts {
                 }
             });
 
+            if (!options.disableSizeLimit && file && file.text && !hasTypeScriptFileExtension(file.fileName)) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: looks like we do not need this in the program building, since we are doing it in the server.

File: src/compiler/commandLineParser.ts [hunk 18]
```diff
@@ -723,10 +727,22 @@ namespace ts {
                 const supportedExtensions = getSupportedExtensions(options);
                 Debug.assert(indexOf(supportedExtensions, ".ts") < indexOf(supportedExtensions, ".d.ts"), "Changed priority of extensions to pick");
 
+                const potentialFiles: string[] = [];
+                if (host.readDirectoryWithMultipleExtensions) {
+                    addRange(potentialFiles, host.readDirectoryWithMultipleExtensions(basePath, supportedExtensions, exclude));
```
Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: I believe this already handled by @riknoll in https://github.com/Microsoft/TypeScript/pull/8841. can we make sure we are doing one approach for both PRs.

Comment (mhegazy) [inline src/compiler/commandLineParser.ts]: possibly split this in a separate PR.

File: src/server/session.ts [hunk 19]
```diff
@@ -335,7 +335,7 @@ namespace ts.server {
         private getTypeDefinition(line: number, offset: number, fileName: string): protocol.FileSpan[] {
             const file = ts.normalizePath(fileName);
             const project = this.projectService.getProjectForFile(file);
-            if (!project) {
+            if (!project || project.languageServiceDiabled) {
```
Comment (mhegazy) [inline src/server/session.ts]: can we instead make getProjectForFile return undefined?

**Comment by mhegazy** [comment]
We need a unit test for this, now that we have unit tests :)

**Comment by mhegazy** [comment]
@vladima any comments?

File: src/compiler/commandLineParser.ts (hunk 18 above)
Comment (zhengbli) [inline src/compiler/commandLineParser.ts]: Sure.

File: src/server/session.ts (hunk 19 above)
Comment (zhengbli) [inline src/server/session.ts]: We still need the `getProjectInfo` call to work, because the client needs to know what files are in the project and is the language service of the project disabled or not. So it's not good idea to never return the project when the language service was disabled.

**Comment by zhengbli** [comment]
@mhegazy updated with a test

File: src/server/editorServices.ts [hunk 20]
```diff
@@ -1299,7 +1416,43 @@ namespace ts.server {
                     return errors;
                 }
                 else {
-                    const oldFileNames = project.compilerService.host.roots.map(info => info.fileName);
+                    if (this.exceedTotalNonTsFileSizeLimit(projectOptions.files) && projectOptions.compilerOptions && !projectOptions.compilerOptions.disableSizeLimit) {
```
Comment (mhegazy) [inline src/server/editorServices.ts]: move this `!projectOptions.compilerOptions.disableSizeLimit` to the beginning.

**Comment by mhegazy** [comment]
👍
